### DISCUSSION TITLE
How to mount multiple `InMemory` backends

### DISCUSSION THREAD

**Discussion by lvcabral** [zen-fs/core#143] [General]
Doing the configuration like this, we end up with only one mount `cachefs:`
![image](https://github.com/user-attachments/assets/6847aa4f-d6c2-4022-b47e-d322e4461be9)
How can I create separate inMemory mounts?

**Comment by james-pre** [comment]
@lvcabral, thanks for reaching out. You will need to wait for `configure` to resolve before mounts are loaded. If you try to use `fs` before then, you will just have the default `InMemory` mounted on `/`.

A suggestion, using multiple different `InMemory` backends mounted at different places does not include any meaningful benefits. If may be easier to just use the default `InMemory`.

When attempting to reproduce this locally, everything works correctly:
```js
import { configure, fs, InMemory } from '@zenfs/core';

await configure({
	mounts: {
		'common:': InMemory,
		'pkg:': InMemory,
		'tmp:': InMemory,
		'cachefs:': InMemory,
	},
});

console.log(...fs.mounts.keys());
```

```sh
$ node index.js
/ /common: /pkg: /tmp: /cachefs:
```

**Reply by lvcabral** [reply]
What you mean by "wait for configure to resolve before mounts are loaded"??
I got your suggestion, but I do have specific benefit in my use case to have separate InMemory mounts as I am replacing an existing solution that relies on separate "volumes" and would be a big change if I can't have it like that.

**Reply by lvcabral** [reply]
In you test you checkted the mount keys, I checked the actual file system using `readdirSync()`

![image](https://github.com/user-attachments/assets/e7b21c67-09a5-4ae6-9e0a-d99c4e2b1395)

**Reply by james-pre** [reply]
> What you mean by "wait for configure to resolve before mounts are loaded"??

`configure` is not synchronous you will need to `await` it (like in my example) or use `.then` (`configure(...).then(() => { fs.something(...) })`). 

> I got your suggestion ...

Yeah, if your use case needs it, there is very little performance overhead to using multiple `InMemory` mounts.

**Reply by lvcabral** [reply]
I was already awaiting the configure in my code.
I believe there should be a way to create new instances of InMemory for each volume, so I can access those independently with specific URLs:
`pkg:/source/main.brs` or `tmp:/images/mypic.png` using the same fs

**Reply by lvcabral** [reply]
Of course the contents of each volume/root folder should be different

**Reply by james-pre** [reply]
> In you[r] test you checked the mount keys, I checked the actual file system using `readdirSync()`

I just got this as well, it was a race condition in `configure` since `mount` creates new directories. I'm surprised it didn't get caught until now. https://github.com/zen-fs/core/commit/e40bf4c27af6f0a541a260857307a67f42f5adc2 should fix it.

I'm adding tests to prevent a regression now.

**Reply by lvcabral** [reply]
Cool, thanks for the quick feedback
